### ISSUE TITLE
Debug

### ISSUE THREAD

**Issue by Karaulov** [alliedmodders/amxmodx#794]
Please add debug to AMXX!! 

For example output to server log all messages from plugins:

Example:
[TIME] Plugin 'testdebug' tried to send message with name (or id) 'SayText'.

This will be useful for find plugins which send bad messages, or cause 'overflow' problem.
**Many plugins not have source code!**

**Comment by OciXCrom** [comment]
No, all plugins have source code. If you don't have a source code, it's your fault you put that plugin in your server in the first place. It's like drinking something without knowing what it's made of - it can kill you.

**Comment by afwn90cj93201nixr2e1re** [comment]
It's not related to src at all, he talking about logging, like it's implemented in metamod.

But logging it's blocking I/O, so, it's not nice.

So, you can add it by ur self.

**Comment by Karaulov** [comment]
> It's not related to src at all, he talking about logging, like it's implemented in metamod.
> 
> But logging it's blocking I/O, so, it's not nice.
> 
> So, you can add it by ur self.
@afwn90cj93201nixr2e1re, but in metamod this is added, why amxx is worse? :)

@OciXCrom, no many plugins without source code (antihack, private, paid)

**Comment by OciXCrom** [comment]
Read the AMXX licence. Sharing plugins without the source code is not allowed, so there's no point in even discussing this.

And why would "antihack" plugins be without source?...

**Comment by Karaulov** [comment]
@OciXCrom, I do not know why no one read this license :rofl: 

Many anti-hacks do not have source code, private and paid plugins, too.
Antihacks uses private method to cheat detection.
Paid... you know why.

**Comment by Karaulov** [comment]
@OciXCrom This is not a reason to not add debugging to AMXX.

**Comment by Karaulov** [comment]
@OciXCrom If author of AMMX present full plugin decompiler with support all protection, all plugins will be have sourcecode))))

**Comment by xLeviNx** [comment]
Can we close this thread? If you don't have access to the source code, that's on you, as per the license, you should have it and if you can't debug the source, then that's also on you. This is not an AMXX issue at all.

**Comment by Karaulov** [comment]
Metamod have source code but also have debug, why amxmodx no have debug?

**Comment by Karaulov** [comment]
@xLeviNx sourcecode has only plugin author, i use only amxx file

**Comment by Karaulov** [comment]
@xLeviNx even with source code, only  AMXX can be debug all plugins

**Comment by Arkshine** [comment]
Please don't spam. Use the edit button if you want to add additional information.
Let's stay on the subject. Source code is not relevant here.

**Comment by Karaulov** [comment]
If need really source code, need disable support compiled amxx plugins and add support sma files.

**Comment by OciXCrom** [comment]
Settings aside (ignoring) the reasonings you provided, such a feature can indeed be useful in some cases, especially since the smallest mistake when using `message_*` can crash the entire server.

**Comment by Karaulov** [comment]
I also suggest to adding amx_github command to generate filled template for github issue(os/game version/etc, as in issue template)

**Comment by DarthMan** [comment]
> 
> 
> Read the AMXX licence. Sharing plugins without the source code is not allowed, so there's no point in even discussing this.
> 
> And why would "antihack" plugins be without source?...

Almost all anti-hack plug-ins that were made, they almost all come with no source code, no matter if metamod module or amxx plug-in. They want to prevent people from seeing how it was coded probably they most likely also include protection against Lysis and some other things. Sure, if you're good at assembly and reverse-engineering, you can easely recreate the plug-in. But not everyone is good at such, there are many good coders who know nothing about IDA but are very good at C++ par example. The reason for that is because if they see the source code, they might be able to find methods to cheat and bypass such protections. But it must also be noted that using such plug-ins most likely means that the server is steam + non-steam. I'm not a lawyer, and I must admit, I'm not for, nor am I against people using dproto, but, as per Allied Modders rules, dproto is considered illegal. And I also understand that there are countries where CS 1.6 is expensive, the equivalent of say 30 US dollars. Countries such as Argentina are where such games are expensive. But still, No Steam = No support here. Such protection plug-ins are useless if the server is installed via steamcmd and is legit, meaning steam-only clients to be allowed. Also, personally, I never liked using dproto because peole can easely fake their authid's, they can change their ip's and even their names, so there's no possible way to truly ban a non-steam player. Also, banning by ip is not a reliable method, as the ip can be dynamic and it can be allocated to another person at any given time, when a disconnection is made and a new connection is made from another place. Even if the ip is static, there can be cases as it happened to me long ago when I had internet from LandNet, when at some point my static ip had to be changed, as they called me and gave a new ip. And my old ip was then later assigned to another person, they said they are going to do that. LandNet neighborhood / district internet company, was then acquired and now my internet is from a public cable company, and I have dynamic ip that changes whenever I connect using PPPOE connection by authenticating using a username and password. In conclusion, this issue can be closed. Maybe just a macro to AMXX, say AMXDEBUG that, if enabled, when manually compiling the source, AMXX will have debug mod enabled.
